packages/pubsub/src/subscription.js

```diff
@@ -592,30 +365,249 @@ Subscription.prototype.decorateMessage_ = function(message) {
+      }
+    }
+  });
+
+  this.on('removeListener', function(event) {
+    if (event === 'message' && --self.messageListeners === 0) {
+      self.closeConnection_();
+    }
+  });
+};
+
+/**
+ *
+ */
+Subscription.prototype.createMessage_ = function(data) {
+  var message = new Message(self, data);
+  this.messages_.add(message);
+  return message;
+};
+
+/**
+ *
+ */
+Subscription.prototype.modifyAckDeadline_ = function(messages, callback) {
+  var self = this;
+  var reqOpts;
+
+  if (this.connection) {
+    reqOpts = messages.reduce(function(reqOpts, message) {
+      message.modifyDeadlineAckIds.push(message.ackId);
+      message.modifyDeadlineSeconds.push(message.deadline);
+      return reqOpts;
+    }, {
+      modifyDeadlineAckIds: [],
+      modifyDeadlineSeconds: []
+    });
+
+    this.connection.write(reqOpts);
+    setImmediate(onComplete);
+    return;
+  }
+
+  var requests = groupByDeadline(messages).map(function(reqOpts) {
+    return self.request({
+      client: 'subscriberClient',
+      method: 'modifyAckDeadline',
+      reqOpts: reqOpts
+    }).catch(function(err) {
+      self.emit('error', err);
+    });
+  });
+
+  Promise.all(requests)
+    .then(onComplete.bind(null, null), onComplete);
+
+  function groupByDeadline(messages) {
+    var requests = new Map();
+
+    messages.forEach(function(message) {
+      if (!requests.has(message.deadline)) {
+        requests.set(message.deadline, {
+          subscription: self.name,
+          ackIds: [],
+          ackDeadlineSeconds: message.deadline
+        });
+      }
+
+      var request = requests.get(message.deadline);
+      request.ackIds.push(message.ackId);
+    });
+
+    return Array.from(requests.values());
+  }
+
+  function onComplete(err, resp) {
+    messages.forEach(function(message) {
+      if (message.deadline === 0) {
+        self.messages_.delete(message);
+      }
+    });
+
+    if (self.connection.isPaused()) {
+      if (self.messages_.size < self.flowControl.maxMessages) {
+        self.connection.resume();
+      }
+    }
+
+    callback(err, resp);
+  }
+};
+
+/**
+ * @param {object} config - The push config.
+ * @param {string} config.pushEndpoint
+ * @param {object} config.attributes
+ */
+Subscription.prototype.modifyPushConfig = function(config, gaxOpts, callback) {
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
+
+  var reqOpts = {
+    subscription: this.name,
+    pushConfig: config
+  };
+
+  this.request({
+    client: 'subscriberClient',
+    method: 'modifyPushConfig',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
+
+/**
+ * @private
+ */
+Subscription.prototype.openConnection_ = function() {
+  var self = this;
+
+  var reqOpts = {
+    subscription: this.name,
+    streamAckDeadlineSeconds: this.ackDeadline
+  };
+
+  this.request({
+    client: 'subscriberClient',
+    method: 'streamingPull',
+    returnFn: true,
+    reqOpts: {
+      highWaterMark: self.flowControl.highWaterMark
```

Comment:
Okay, I am tracking the issue now. (And yeah, you had this discussion with @jmuk, which explains why I lacked the context.)

@callmehiphop I am pretty sure that @jganetsk is correct here. Essentially, this is what we actually want: say that we are getting more messages from the stream than we can handle. If there is a message and no available processes to handle it (or flow control maximums are hit), we basically want to `nack` it immediately and then just throw it away.

What this does in practice is that it makes _the Pub/Sub service_ into the buffering agent. Rather than having the client buffer, the server will.

This is an important distinction because there is a pretty good chance that the single client that is running is not actually the only client on the playground. Straight-up `nack`ing the message means that the server can immediately pass the message to someone else.